src/reanimated2/animation/decay.ts

```diff
@@ -0,0 +1,134 @@
+import { defineAnimation } from './util';
+import {
+  Animation,
+  AnimationCallback,
+  PrimitiveValue,
+  Timestamp,
+} from './commonTypes';
+import { Platform } from 'react-native';
+
+interface DecayConfig {
+  [key: string]: any;
+  deceleration?: number;
+  velocityFactor?: number;
+  clamp?: number[];
+  velocity?: number;
+}
+
+export interface DecayAnimation extends Animation<DecayAnimation> {
+  lastTimestamp: Timestamp;
+  startTimestamp: Timestamp;
+  initialVelocity: number;
+  velocity: number;
+  current: PrimitiveValue;
+}
+
+export interface InnerDecayAnimation extends Omit<DecayAnimation, 'current'> {
+  current: number;
+}
+
+export function withDecay(
+  userConfig: DecayConfig,
+  callback?: AnimationCallback
+): Animation<DecayAnimation> {
+  'worklet';
+
+  return defineAnimation<DecayAnimation>(0, () => {
+    'worklet';
+    const config: Required<DecayConfig> = {
+      deceleration: 0.998,
+      velocityFactor: Platform.OS !== 'web' ? 1 : 1000,
+      clamp: [],
+      velocity: 0,
+    };
+    if (userConfig) {
+      Object.keys(userConfig).forEach((key) => (config[key] = userConfig[key]));
+    }
+
+    const VELOCITY_EPS = Platform.OS !== 'web' ? 1 : 1 / 20;
+    const SLOPE_FACTOR = 0.1;
+
+    function decay(animation: InnerDecayAnimation, now: number): boolean {
+      const {
+        lastTimestamp,
+        startTimestamp,
+        initialVelocity,
+        current,
+        velocity,
+      } = animation;
+
+      const deltaTime = Math.min(now - lastTimestamp, 64);
+      const v =
+        velocity *
+        Math.exp(
+          -(1 - config.deceleration) * (now - startTimestamp) * SLOPE_FACTOR
+        );
+      animation.current =
+        current + (v * config.velocityFactor * deltaTime) / 1000; // /1000 because time is in ms not in s
+      animation.velocity = v;
+      animation.lastTimestamp = now;
+
+      if (config.clamp) {
+        if (initialVelocity < 0 && animation.current <= config.clamp[0]) {
+          animation.current = config.clamp[0];
+          return true;
+        } else if (
+          initialVelocity > 0 &&
+          animation.current >= config.clamp[1]
+        ) {
+          animation.current = config.clamp[1];
+          return true;
+        }
+      }
+
+      if (Math.abs(v) < VELOCITY_EPS) {
+        return true;
+      }
+      return false;
+    }
+
+    function validateConfig(): void {
+      if (config.clamp) {
+        if (Array.isArray(config.clamp)) {
+          if (config.clamp.length !== 2) {
+            console.error(
+              `clamp array must contain 2 items but is given ${config.clamp.length}`
+            );
+          }
+        } else {
+          console.error(
+            `config.clamp must be an array but is ${typeof config.clamp}`
+          );
+        }
```

Comment:
Maybe not in scope of this PR but we could invert those ifs to flatten this a bit:
```js
if (config.clamp) {
	if (!Array.isArray(config.clamp)){
		throw //typeof error
	}
	if(config.clamp.length !== 2){
		throw //length error
	}
}
```